hadoop-hdds/docs/content/design/overwrite-key-only-if-unchanged.md

```diff
@@ -0,0 +1,149 @@
+5. The client continues to write the data as usual.
+6. On commit key, the client does not need to send the overwriteExpectedUpdateID again, as the open key contains it.
+7. On OM, on commit key, it validates the key still exists with the given key name and its updateID is unchanged. If so the key is committed, otherwise an error is returned to the client.
+
+## Changes Required
+
+In order to enable the above steps on Ozone, several small changes are needed.
+
+### Wire Protocol
+
+1. The overwriteExpectedUpdateID needs to be added to the KeyInfo protobuf object so it can be stored in the openKey table.
+2. The overwriteExpectedUpdateID needs to be added to the keyArgs protobuf object, which is passed from the client to OM when creating a key.
+
+No new messages need to be defined.
+
+### On OM
+
+No new OM handlers are needed. The existing OpenKey and CommitKey handlers will receive the new overwriteExpectedUpdateID and perform the checked.
+
+No new locks are needed on OM. As part of the openKey and commitKey, there are existing locks taken to ensure the key open / commit is atomic. The new checks are performed under those locks, and come down to a couple of long comparisons, so add negligible overhead.
+
+### On The Client
+
+ 1. We need to allow the updateID of an existing key to be accessible when an existing details are read, by adding it to OzoneKey and OzoneKeyDetails. There are internal object changes and do no impact any APIs.
+ 2. To pass the overwriteExpectedUpdateID to OM on key open, it would be possible to overload the existing OzoneBucket.createKey() method, which already has several overloaded versions, or create a new explicit method on Ozone bucket called replaceKeyIfUnchanged, passing either the OzoneKeyDetails of the existing key (which includes the key name and existing updateID, or by passing the key name and updateID explicitly, eg:
+ 
+ ```
+ public OzoneOutputStream replaceKeyIfUnchanged(OzoneKeyDetails keyToOverwrite, ReplicationConfig replicationConfigOfNewKey)
+      throws IOException 
+	  
+// Alternatively or additionally
+
+ public OzoneOutputStream replaceKeyIfUnchanged(String volumeName, String bucketName, String keyName, long size, long expectedUpdateID, ReplicationConfig replicationConfigOfNewKey)
+      throws IOException 
+
+	  
+ ```
+This specification is roughly in line with the exiting createKey method:
+
+```
+  public OzoneOutputStream createKey(
+      String volumeName, String bucketName, String keyName, long size,
+      ReplicationConfig replicationConfig,
+      Map<String, String> metadata)
+```
+
+An alternative, is to create a new overloaded createKey:
+
+```
+  public OzoneOutputStream createKey(
+      String volumeName, String bucketName, String keyName, long size,
+      ReplicationConfig replicationConfig, long expectedUpdateID)
+```
+
+Note the omission of the metaData map, as the intention of this API is to copy that from what already exisits on the server.
```

Comment:
The existing code, on overwriting a key, expects new meta-data to be passed. It does not copy what is there on the server to the new key. This is, for reasons unknown to me, different from ACLs, which are copied.

The omission of the metadata map, is therefore to remove the ability to pass new metadata, but also to remove the need to copy the old metadata over, making it simpler to use.

I can of course allow the metadata map to be supplied as with other createKey operations, but then it makes this new API perhaps harder to use, and easier to make a mistake with.

If all you want to do, is change the replicationType of a key, then you need to ensure you copy the old metadata map client side. Having a null metadata map is a valid option too, so you cannot tell server side what the intent was if it is missing.

I am happy to go either way on this, but the above is my reasoning for why I designed it this way.

Reply:
> The existing code, on overwriting a key, expects new meta-data to be passed. It does not copy what is there on the server to the new key. This is, for reasons unknown to me, different from ACLs, which are copied.

Yeah I think this ACL handling is wrong too, I don't think we need to worry about that part here.

> The omission of the metadata map, is therefore to remove the ability to pass new metadata, but also to remove the need to copy the old metadata over, making it simpler to use.

This ability is defined by the protos and the server, not the wrapper method we provide at the client. Since we are re-using the existing create and commit requests which support metadata alterations, the rewrite API that re-uses these requests would also end up supporting it. To "not support it" the server has to actually reject requests that alter these fields. This is why I said it is actually easier to let metadata updates with create/commit happen as part of the rewrite.

> To keep the scope as small as possible, I am not considering metadata only changes at the current time.

Agreed. I was only referring to cases where metadata happens to be changed as part of the rewrite, not suggesting a separate operation that leaves all the blocks and does a metadata only swap. Actually that operation would be a single transaction so I think it just works already.

Reply:
I am happy to implement this inline with the current createKey when overwrite - which is the meta data will have to be explicitly copied from the existing key into the new key by the client. That keeps things consistent.

Reply:
@errose28 is there anything still outstanding here?

Reply:
I think we are good here. Sounds like we are in agreement that metadata copying will work as usual from the server and API perspective, but the client's methods don't need to expose this functionality right now.